src/conversation_engine/engine.py

```diff
@@ -0,0 +1,409 @@
+        return self._state
+
+    @property
+    def target(self):
+        return self._target
+
+    @property
+    def missing_field(self):
+        return self._missing_field
+
+    @property
+    def current_semantics(self):
+        return self._current_semantics
+
+    def wait_for_user(self, target, missing_field):
+        rospy.loginfo("ConversationState: {old} -> {new}. Target='{t}', missing_field='{mf}'"
+                      .format(old=self._state, new=ConversationState.WAIT_FOR_USER,
+                              t=target, mf=missing_field))
+        self._state = ConversationState.WAIT_FOR_USER
+        self._target = target
+        self._missing_field = missing_field
+
+    def wait_for_robot(self):
+        rospy.loginfo("ConversationState: {old} -> {new}".format(old=self._state, new=ConversationState.WAIT_FOR_ROBOT))
+        self._state = ConversationState.WAIT_FOR_ROBOT
+        self._target = None
+        self._missing_field = None
+
+    def aborting(self, timeout, timeout_callback):
+        """Set the state to ABORTING. If the action server hasn't aborted the action after the given timeout,
+        call the callback to deal with that
+        :param timeout duration to wait before killing
+        :type timeout rospy.Duration
+        :param timeout_callback callback is called when the aborting takes too long
+        :type callable(event: rospy.TimerEvent)"""
+        rospy.loginfo("ConversationState: {old} -> {new}".format(old=self._state, new=ConversationState.ABORTING))
+        self._state = ConversationState.ABORTING
+
+        rospy.Timer(timeout,
+                    timeout_callback,
+                    oneshot=True)
+
+    def initialize_semantics(self, semantics):
+        self._current_semantics = semantics
+        rospy.loginfo("Initialized semantics: {}".format(self._current_semantics))
+
+    def update_semantics(self, semantics, missing_field_path):
+        # I assume semantics is the exact information requested and supposed to go in place of the field indicated by
+        # the missing information path
+
+        # fill semantics in existing semantics
+        path_list = missing_field_path.split('.')
+        action_index = int(path_list[0].strip('actions[').rstrip(']'))
+        fields = path_list[1:]
+
+        elem = self._current_semantics['actions'][action_index]
+        for field in fields:
+            try:
+                elem = elem[field]
+            except KeyError:
+                elem[field] = semantics
+
+
+        rospy.loginfo("Updated semantics: {}".format(self._current_semantics))
+        return
+
+
+class ConversationEngine(object):
+    def __init__(self, robot_name):
+        self._knowledge = knowledge_loader.load_knowledge('challenge_open')
+
+        self._state = ConversationState()
+
+        self._action_client = Client(robot_name)
+
+        self._parser = cfgparser.CFGParser.fromstring(self._knowledge.grammar)
+        self._robot_name = robot_name
```

Comment:
This is because the action_server's client does ```action_name = "/" + robot_name + "/action_server/task"```. That is the only reason AFAIK

Otherwise we could just use ROS namespacing